### PR TITLE
[module_system] traverse and parse all dependent modules

### PR DESCRIPTION
**Pull request by rayshih**
```
% ./build/exec/violet check AModule
parse module ./example/AModule.vt
parse module ./example/Base.vt
parse module ./example/Lib.vt
parse module ./example/Core.vt
Nat : U
zero : Nat
suc : Nat → Nat
id : {A : U} → (x : A) → A
main : Nat
```

------

Next step: topological sort and compile

------

Few topics to discuss:

1. the mapping from module name to filepath, including the root of the "package"
2. Error handling beyond the one file typechecker/compiler
3. Scope of each module -> simple one could be exposing all top level decls to the only one global ctx/env

### CODE REVIEW

File: example/Base.vt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module Base
+
+import Core
```
Comment (dannypsnl) [inline example/Base.vt:3]: eof

File: example/Core.vt [hunk 2]
```diff
@@ -0,0 +1 @@
+module Core
```
Comment (dannypsnl) [inline example/Core.vt:1]: eof

File: example/Lib.vt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+module Lib
+
+import Base
+import Core
```
Comment (dannypsnl) [inline example/Lib.vt:4]: eof

**Review by dannypsnl — CHANGES_REQUESTED** [review]

**Comment by rayshih** [comment]
> 1. the mapping from module name to filepath, including the root of the "package"

The essential question is: given a module name, where to lookup the corresponding file. And there are 2 smaller questions:

(Assuming the module to filepath mapping is mostly replacing `.` to `/`
a. should we introduce a new argument for specifying the root directory of all modules?
b. if we extend this question, then the next question is do we want to introduce a package system? as in, every package has its own root directory.

> 2. Error handling beyond the one file typechecker/compiler

The current error handling is limited to one file, and I haven't figured out a way to put cross module errors. I think the problem is: `App` monad is only about one file. Should we extend it to multiple files? or do you have other ideas?


> 3. Scope of each module -> simple one could be exposing all top level decls to the only one global ctx/env

Each module has its own scope of top decls, to run a typecheck of a module A which depends on module B, the exported decls from B needs to also be in the scope while typechecking module A. The easiest solution is

a. assume all top decls are exported.
b. bring all dependent module top decls into the current scope.

But there are problem's like if there are 2 decls from different modules having the same name, which one to take?

Alternatively we can create a lookup map. But in any case, we will need to have a resolution method if there is naming conflict

**Comment by dannypsnl** [comment]
> 1. the mapping from module name to filepath, including the root of the "package"

Both answers are yes, we should introduce package concept, which is a collection of modules. A package is a root of files, and should have a meta file to describe how to deal with these stuff.

> 2. Error handling beyond the one file typechecker/compiler

I am not quite sure for the details now, but I have some constraints in mind that you can follow
1. The module will not get check, if its dependencies get failed
2. But each dependencies are independent 

> 3. Scope of each module -> simple one could be exposing all top level decls to the only one global ctx/env

Keep the simple solution for now since name resolution is a complicated thing, we will need an isolated issue to track it, you can open one and describe the problem.

**Comment by rayshih** [comment]
let's land this PR first and I'll continue on the toposort and then the ctx/env propagation

File: src/Main.idr [hunk 4]
```diff
@@ -25,10 +30,49 @@ checkMod filename source raw = do
...
+			(raw, source) <- 
+				parseModuleFile $ moduleNameToFilepath root moduleName
+			let imports = getImports raw
+			let acc' = insert moduleName (raw, imports, source) acc
+			parseDepModules root acc' (rest ++ Data.SortedSet.toList imports)
+
+loadModuleFile : (PrimIO e, FileIO (IOError :: e)) => String -> App e CheckState
+loadModuleFile filename = do
+	modules <- parseDepModules "./example" empty $ singleton filename
+	-- (raw, source) <- parseModuleFile filename
```
Comment (dannypsnl) [inline src/Main.idr:71]: Maybe we don't need this unused code?

File: src/Main.idr [hunk 5]
```diff
@@ -25,10 +30,49 @@ checkMod filename source raw = do
...
+		Nothing => do
+			(raw, source) <- 
+				parseModuleFile $ moduleNameToFilepath root moduleName
+			let imports = getImports raw
+			let acc' = insert moduleName (raw, imports, source) acc
+			parseDepModules root acc' (rest ++ Data.SortedSet.toList imports)
+
+loadModuleFile : (PrimIO e, FileIO (IOError :: e)) => String -> App e CheckState
+loadModuleFile filename = do
+	modules <- parseDepModules "./example" empty $ singleton filename
```
Comment (dannypsnl) [inline src/Main.idr:70]: Please mark this `"./example"` hard coding with TODO or FIXME

**Review by dannypsnl — CHANGES_REQUESTED** [review]

**Review by dannypsnl — APPROVED** [review]

**Pull request merged**